tls/s2n_cipher_suites.c

```diff
@@ -1168,11 +1171,158 @@ static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, u
+        /* If the kex is not supported continue to the next candidate */
+        bool kex_supported = false;
+        if (s2n_result_is_error(s2n_kex_supported(*potential_match, conn, &kex_supported))) {
+            return S2N_RESULT_OK;
+        }
+
+        if (!kex_supported) {
+            return S2N_RESULT_OK;
+        }
+
+        /* If the kex is not configured correctly continue to the next candidate */
+        if (s2n_result_is_error(s2n_configure_kex(*potential_match, conn))) {
+            return S2N_RESULT_OK;
+        }
+    }
+
+    /**
+    *= https://tools.ietf.org/rfc/rfc8446#section-4.2.11
+    *# The server MUST ensure that it selects a compatible PSK
+    *# (if any) and cipher suite.
+    **/
+    if (conn->psk_params.chosen_psk != NULL) {
+        if ((*potential_match)->prf_alg != conn->psk_params.chosen_psk->hmac_alg) {
+            return S2N_RESULT_OK;
+        }
+    }
+
+    *match_is_valid = true;
+    return S2N_RESULT_OK;
+}
+
+/* If both server and client signaled support for ChaCha20 boosting, then the corresponding ChaCha20 cipher suite is set. Else, NULL is set. */
+static S2N_RESULT s2n_get_boosted_cipher_suite(struct s2n_connection *conn, struct s2n_stuffer wire_stuffer, uint32_t count, uint32_t cipher_suite_len, struct s2n_cipher_suite **boosted_cipher_suite) {
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(boosted_cipher_suite);
+
+    const struct s2n_security_policy *security_policy;
+    RESULT_GUARD_POSIX(s2n_connection_get_security_policy(conn, &security_policy));
+
+    /* First check if the server has enabled chacha20 boosting */
+    *boosted_cipher_suite = NULL;
+    if (!security_policy->cipher_preferences->allow_chacha20_boosting) {
+        return S2N_RESULT_OK;
+    }
+
+    /* Next, check if the client has a chacha20 cipher as its most preferred cipher suite */
+    uint8_t clients_first_cipher_iana[S2N_SSLv2_CIPHER_SUITE_LEN] = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_read_bytes(&wire_stuffer, clients_first_cipher_iana, cipher_suite_len));
+
+    bool is_chacha20_cipher_suite = false;
+    RESULT_GUARD(s2n_iana_is_a_chacha20_cipher_suite(clients_first_cipher_iana, &is_chacha20_cipher_suite));
+    if (!is_chacha20_cipher_suite) { 
+        return S2N_RESULT_OK;
+    }
+
+    /* Check if the server has the boosted chacha20 cipher suite in its cipher preferences */
+    bool server_supports_boosted_cipher = false;
+    RESULT_GUARD(s2n_cipher_preferences_contain(security_policy->cipher_preferences, clients_first_cipher_iana, &server_supports_boosted_cipher));
+    if (!server_supports_boosted_cipher) {
+        return S2N_RESULT_OK;
+    }
+
+    /* We know the client is boosting a supported chacha20 cipher suite; retrieve the cipher suite from the iana */
+    struct s2n_cipher_suite *chacha20_cipher_suite = NULL;
+    RESULT_GUARD(s2n_cipher_suite_from_iana(clients_first_cipher_iana, &chacha20_cipher_suite));
+    RESULT_ENSURE_REF(chacha20_cipher_suite);
```

Comment:
We trigger chacha20 boosting if any chacha20 cipher suite is first in the list. But do we choose the cipher suite that's first in the client's list, or do we choose the chacha20 cipher suite that's first in the local server list? There's more in a TLS1.2 cipher suite than just an encryption alg.

Reply:
The first cipher on the client's list might also fail the s2n_validate_cipher_suite_match check below. Imagine if the client prefers ECDSA auth but we only have RSA.

Reply:
It should be cipher suite that's first in the clients list. If the validation fails, then `boosted_cipher_suite` is never set and we don't proceed with the boosting logic

Reply:
Why would it be the cipher suite that's first in the client's list?

Reply:
Because that's how client's signal boosting. If the server has chacha20 boosting support enabled and the client has a chacha20 cipher suite first in its list, then we boost that cipher suite (provided it is also supported by the server).

Reply:
Let's look at an example:
* Client wire list: "ECDHE-RSA-CHACHA20-POLY1305, ECDHE-ECDSA-CHACHA20-POLY1305, TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA38"
* Server preferences: "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA38, ECDHE-ECDSA-CHACHA20-POLY1305"

The client clearly prefers chacha20. It likely doesn't even care about RSA vs ECDSA, it just had to pick an ordering. The server does care about RSA vs ECDSA because it likely only has one certificate, in this case ECDSA.

Your code would choose TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA38. I would expect the outcome our chacha20 boosting customers would want is that we choose ECDHE-ECDSA-CHACHA20-POLY1305.

Reply:
This change would actually choose `ECDHE-RSA-CHACHA20-POLY1305` not `TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA38`.

Reply:
Then you have a more serious problem, because that's not even on the server's preference list so should never be an option.

Reply:
I'm confused, `ECDHE-RSA-CHACHA20-POLY1305` is on the server preference list in your example? 
1. Client placed `ECDHE-RSA-CHACHA20-POLY1305` as most preferred
2. (context is that the server has chacha20 flag toggled on)
3. Server has `ECDHE-ECDSA-CHACHA20-POLY1305` in it's cipher preference list.
4. If `ECDHE-ECDSA-CHACHA20-POLY1305` is valid for the connection, then it is negotiated. Else, fall back to server-preference negotiation only.

Reply:
As discussed offline, the behaviour for chacha20 boosting should be:

1. The server toggled on `allow_chacha20_boosting`
2. The client has a chacha20 cipher suite first in its list
3. The server will then try to negotiate a chacha20 ciphersuite according to server preferences
4. If step 3 succeeds, then finish. Else, fall back to the server-side only preference negotiation